Associate with files and URIs on windows

Susko3 opened the pull request:
- Supersedes #26658 
- Closes #14824

## Fixed problems from the previous PR

- URL associations are implemented
- added localisable file descriptions
- added file icons
- registry keys are properly disposed
- implementation is testable in a reasonable way
- `SHChangeNotify` is now called in a non-blocking way
- fixed stable registry keys taking priority (this is handled by the windows explorer -- it'll ask the user (once) in which program to open)

## Problems with this PR

- [x] (Fixed) associations are re-created every time on startup (it's not great, but it probably doesn't matter that much)
- [x] (Fixed) associations clash with [`getStableInstallPathFromRegistry()`](https://github.com/ppy/osu/blob/2bac09ee00d1e809e38aca97350e4853b2ebf09f/osu.Desktop/OsuGameDesktop.cs#L87-L91)
  - if stable uses `HKEY_LOCAL_MACHINE`, maybe we could use that there instead. since this PR saves to `HKEY_CURRENT_USER`
- [ ] localised file association descriptions are never written

## Testing

~~Can be done trough the added test scene. Only manual testing is available.~~

~~Best way to test in the test scene is to install associations, then try the open/present beatmap/url. You can also uninstall and then check that associations don't work.~~

~~For now, uninstalling is only available trough tests and it's not possible to uninstall from the game.~~

Can be tested by running static methods in `WindowsAssociationManager`.

I have tested a full deploy+install on a test PC and it's working perfectly!

Susko3 committed:
Add way to associate with files and URIs on Windows

Susko3 committed:
Associate on startup

Susko3 committed:
Only associate on a deployed build

Helps to reduce clutter when developing

Susko3 committed:
Uninstall associations when uninstalling from squirrel

Susko3 commented:
Hmm, I'm not sure how to make this testable now that https://github.com/ppy/osu-resources/pull/308 is closed and icons are supposed to be in `osu.Desktop`... Maybe move the current code to `osu.Game.Platform.Windows`, make it more abstract (actual associations are defined somewhere in `osu.Desktop` or `osu.Game.Tests` projects, but `WindowsAssociationManager` knows how to install them).

This could allow us to make eg. `.osz-test` associations for testing, which is much cleaner.

Susko3 committed:
Use cleaner way to specify .exe path

bdach commented:
I'm not sure visual tests for this are useful at this stage. If this is gonna hook into squirrel flows the only actual way to test this is via `osu-deploy`.

I'm kinda uncomfortable with the notion of "visual tests" touching system registry to begin with.

Susko3 committed:
Move `WindowsAssociationManager` to `osu.Desktop`

Susko3 committed:
Copy .ico files when publishing

These icons should appear in end-user installation folder.

Susko3 committed:
Inline constants

Susko3 committed:
Remove tests

Can be tested with
```
dotnet publish -f net6.0 -r win-x64 osu.Desktop -o publish -c Debug
publish\osu!
```

Susko3 committed:
Change to class to satisfy CFS hopefully

bdach commented on osu.Desktop/Windows/Win32Icon.cs:
Don't think this class needs to exist

bdach commented on osu.Desktop/Windows/Icons.cs:
Same for this one. Or, at least, this can/should be a static class holding plain string consts.

bdach commented on osu.Desktop/Windows/WindowsAssociationManager.cs:
```diff
@@ -0,0 +1,265 @@
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+            localisationParameters.ValueChanged += _ => updateDescriptions();
```
It is a bit veiled, but what this ends up implying is that every single change of user language, or even the "show metadata in original language" toggle in settings, _is going to cause writes to the windows registry_. I am not comfortable with this and do not think that is sane.

I think there are two paths I'd rather take here:
- not bother with localising descriptions at all
- localise them, but on a one-shot basis, at point of installation.

bdach replied:
> This relies too heavily on the automatic language selection being correct

What do you mean by "too heavily"? I'm not sure I've known that to fail that much. If there's one place where that should work perfectly it's windows.

> How about updating the localised description once the user selects and confirms their language in the first-run setup wizard?

Eeeeehhhh I don't know. I'd rather not touch the registry outside of squirrel flows.

Susko3 replied:
@ppy/team-client I'm happy with the code and this PR is ready for review. The above conversation remains the only open question.

bdach commented on osu.Desktop/OsuGameDesktop.cs:
```diff
@@ -137,6 +137,9 @@ protected override void LoadComplete()
             if (RuntimeInfo.OS == RuntimeInfo.Platform.Windows)
                 LoadComponentAsync(new GameplayWinKeyBlocker(), Add);
 
+            if (OperatingSystem.IsWindows() && IsDeployedBuild)
+                LoadComponentAsync(new WindowsAssociationManager(), Add);
```
I also don't think that this is right. This shouldn't need to run on every game startup. It should be hooked into squirrel install flows. At the _most_ there should be a button similar stable to manually fix associations when they break, but the game shouldn't be trying to write the same thing to registry every launch and then re-raise notifications about changed associations when nothing really changed.

Susko3 replied:
Agreed. I'll look into making this static and only running once. (Or maybe on every update, in case the icons, etc. are changed.)

bdach commented:
> associations clash with [`getStableInstallPathFromRegistry()`](https://github.com/ppy/osu/blob/2bac09ee00d1e809e38aca97350e4853b2ebf09f/osu.Desktop/OsuGameDesktop.cs#L87-L91)

Can you elaborate on what this means?

Susko3 commented:
> > associations clash with [`getStableInstallPathFromRegistry()`](https://github.com/ppy/osu/blob/2bac09ee00d1e809e38aca97350e4853b2ebf09f/osu.Desktop/OsuGameDesktop.cs#L87-L91)
> 
> Can you elaborate on what this means?

`getStableInstallPathFromRegistry()` queries the current `osu://` URI handler and checks if it is stable. Since lazer will overwrite that handler (for thr current user), the stable install path information will be lost.

File associations for lazer and stable are installed side-by-side, so I'll look into using those for detecting the stable install.

Susko3 committed:
Remove `Win32Icon` class and use plain strings instead

Susko3 committed:
Make `WindowsAssociationManager` `static`

Usages/localisation logic TBD

Susko3 committed:
Install associations as part of initial squirrel install

Susko3 committed:
Remove duplicate try-catch and move `NotifyShellUpdate()` to less hidden place

Susko3 committed:
Move update/install logic into helper

Susko3 committed:
Refactor public methods

Susko3 committed:
Move public methods closer together

Susko3 committed:
Use `Logger.Error`

Susko3 committed:
Refactor ProgID logic so it's more visible

Susko3 committed:
Standardise using declaration

Susko3 committed:
Standardise using declaration

Susko3 committed:
Make things testable via 'Run static method' in Rider

Susko3 committed:
Remove unused param

Susko3 committed:
Merge branch 'master' into windows-file-uri-association

Susko3 committed:
Copy icons to nuget and install package

Don't ask me why this uses the folder for .NET Framework 4.5

Susko3 commented:
I've just checked how this behaves with stable installed. I used a clean slate test PC which hasn't ever had stable or lazer installed.

- stable registers `.osz2`, is this a file format lazer can handle? (a quick search yields that it doesn't)
- stable uses a lowercase `osu! beatmap` as the file description
- stable calls every file `osu! beatmap`
- there is no 'open with' for protocols, so `osu://` is always handled by lazer (HKCU overrides HKLM)
- we can use the `osu!` entry for `getStableInstallPathFromRegistry()`, as that is used exclusively by stable

![comparison](https://github.com/ppy/osu/assets/16479013/3abe4c3a-3402-4b4c-b6f2-4ad9a8162704)

As it currently stands, lazer file associations override stable's, with only lazer appearing in the open with menu:

![only lazer](https://github.com/ppy/osu/assets/16479013/43387b44-9212-43b8-9dd3-f6df0f3a6694)

Luckily, it's a simple fix and can be done either from stable or lazer. The following registry keys need to be added (both `HKLM` and `HKCU` works!):

```reg
[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osk\OpenWithProgIds]
"osu!"=""

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osr\OpenWithProgIds]
"osu!"=""

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osz\OpenWithProgIds]
"osu!"=""

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osz2\OpenWithProgIds]
"osu!"=""
```

![lazer and stable](https://github.com/ppy/osu/assets/16479013/5f2d9f78-bfd5-4527-b5b1-9da0c47cccff)


<details><summary>Registry entries installed by stable (reference)</summary>
<p>

```reg
Windows Registry Editor Version 5.00

; files

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osk]
@="osu!"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osr]
@="osu!"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osz]
@="osu!"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\.osz2]
@="osu!"

; protocol

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu]
@="osu! url handler"
"URL Protocol"=""

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu\DefaultIcon]
@="\"C:\\Users\\User\\AppData\\Local\\osu!\\osu!.exe\",1"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu\shell\open\command]
@="\"C:\\Users\\User\\AppData\\Local\\osu!\\osu!.exe\" \"%1\""

; file handler (only 1)

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu!]
@="osu! beatmap"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu!\DefaultIcon]
@="\"C:\\Users\\User\\AppData\\Local\\osu!\\osu!.exe\",1"

[HKEY_LOCAL_MACHINE\SOFTWARE\Classes\osu!\shell\open\command]
@="\"C:\\Users\\User\\AppData\\Local\\osu!\\osu!.exe\" \"%1\""
```

</p>
</details>

Susko3 committed:
Fix stable install path lookup

`osu` is the `osu://` protocol handler, which gets overriden by lazer.
Instead, use `osu!` which is the stable file handler.

Susko3 committed:
Make `NotifyShellUpdate()` `public` to ease testing

bdach commented on osu.Desktop/Windows/WindowsAssociationManager.cs:
```diff
@@ -0,0 +1,273 @@
+        /// </summary>
+        public const string SHELL_OPEN_COMMAND = @"Shell\Open\Command";
+
+        public static readonly string EXE_PATH = Path.ChangeExtension(typeof(WindowsAssociationManager).Assembly.Location, ".exe").Replace('/', '\\');
```
What's with the replace at the end? This already appears to be correct on my machine...

```csharp
> typeof(Program).Assembly.Location
  "D:\\Open Source\\osu\\osu.Desktop\\bin\\Debug\\net8.0\\osu!.dll"

> Path.ChangeExtension(typeof(Program).Assembly.Location, ".exe")
  "D:\\Open Source\\osu\\osu.Desktop\\bin\\Debug\\net8.0\\osu!.exe"
```

bdach replied:
I dunno that's weird. I'd like to think windows should be able to handle either?

bdach commented on osu.Desktop/Windows/WindowsAssociationManager.cs:
```diff
@@ -0,0 +1,273 @@
+            /// <summary>
+            /// Installs a file extenstion association in accordance with https://learn.microsoft.com/en-us/windows/win32/com/-progid--key
+            /// </summary>
+            public void Install(RegistryKey classes, string exePath)
```
Is there a point to having all these method accept `SOFTWARE_CLASSES` but opened as the first param? Can that just be inlined here?

Susko3 replied:
It's to avoid repeatedly opening and closing the same registry key.

bdach replied:
Is that an issue though? I'm looking at this from a maintainability angle, wherein it's a public method anyone can point a random registry key at. I'd like our interactions with the registry to be very particular and limited in scope.

bdach commented on osu.Desktop/Windows/WindowsAssociationManager.cs:
```diff
@@ -0,0 +1,273 @@
+
+            public void Uninstall(RegistryKey classes)
+            {
+                // importantly, we don't delete the default program entry because some other program could have taken it.
```
Doesn't this mean there'll be a hanging entry there? Is windows going to handle that fine?

Wouldn't it be better to just check if the key can be safely uninstalled (if it has the expected value then it can be removed)?

Susko3 replied:
> Doesn't this mean there'll potentially be a hanging entry there to a nonexistent executable?

Hanging entry to a nonexistent ProgID. [MS recommends to just leave it like that](https://learn.microsoft.com/en-us/windows/win32/shell/fa-file-types#deleting-registry-information-during-uninstallation) (I'll add this link to the inline comment).

> Is windows going to handle that fine?

The only visible issue is that the associated file types will be shown as "osu.File.osz" in explorer. Whereas windows by default would show "OSZ file".

![image](https://github.com/ppy/osu/assets/16479013/f360a46d-323a-4826-84c2-8dcf2f728940)

![image](https://github.com/ppy/osu/assets/16479013/5f1eb13b-3f69-427d-8aa5-acae3c1494a9)

> Wouldn't it be better to just check if the key can be safely uninstalled (if it has the expected value then it can be removed)?

Yeah, better to remove so it doesn't show "osu.File.*" to the user.

peppy commented:
Hope we can get this in the next release.

Susko3 committed:
Refine uninstall logic to account for legacy windows features

Susko3 committed:
Inline `EXE_PATH` usage

Susko3 committed:
Inline `HKCU\Software\Classes` usage

bdach committed:
Privatise registry-related constants

Don't see any reason for them to be public.

bdach commented:
I've tested this and it seems to be behaving pretty well, and getting along with stable pretty amicably even. Did just one small cleanup commit (https://github.com/ppy/osu/pull/27001/commits/87509fbf6efc9e14979a97fdb14b5fc6b56bdf16), looks good to go otherwise.

peppy committed:
Fix typos in xmldoc

peppy reviewed (approved):
Works well for me!
